justfile

```
@@ -0,0 +1,51 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Check if the Slurm snap exists before running a recipe.
+_check_snap_exists:
+  @[ -f slurm_*.snap ] || \
+    { echo "Slurm snap must be built for this recipe to run. Run \`just snap\` to build snap". \
+     && exit 1; }
+
+# Build snap
+snap:
+  snapcraft -v pack
+
+# Apply coding style standards to project
+fmt:
+  tox run -e fmt
+
+# Check project against coding style standards
+lint:
+  tox run -e lint
+
+# Run unit tests
+unit:
+  tox run -e unit
+
+# Run integration tests
+integration: _check_snap_exists
+  #!/usr/bin/env bash
```

Why not add a dependency on `snap` instead?

```suggestion
integration: snap
  #!/usr/bin/env bash
```

---

We can't depend on the `snap` recipe unfortunately. There's a longstanding bug within snapcraft's Python plugin on `core22` that causes it to not be idempotent. The snapcraft will fail on iterative builds of the Slurm snap: https://github.com/canonical/craft-parts/issues/675

`_check_snap_exists` is a way to get around ensuring the snap package to exist to run the integration tests but get around the `craft-parts` bug mentioned above. If we depended on the `snap` recipe, the integration tests will fail if the snap has already been successfully been built once.

You can reproduce the issue by running `just snap` twice:

```shell
just snap
just snap
```

---

Then change the snap recipe to clean before packing? If that's the bug, we would need to do that manually anyways, right?

---

Hmm... I had to check that `snapcraft clean` would cause issues before running a fresh build, but looks like everything is good. I made `snap` depend on `clean` which ensures that we have a fresh build instance each time `just snap` is called. Now the `integration` recipe depends on the `snap` recipe rather than a hidden recipe that checks for the _*.snap_ file.

---

Resolved in https://github.com/charmed-hpc/slurm-snap/pull/51/commits/dc95a3adfa0ece30518a369fbc4f9a5d95686461